jrt/src/com/yahoo/jrt/SslSocket.java

```diff
@@ -0,0 +1,232 @@
+        }
+
+        while (true) {
+            switch (sslEngine.getHandshakeStatus()) {
+                case NOT_HANDSHAKING:
+                    if (hasWrapBufferMoreData()) return HandshakeState.NEED_WRITE;
+                    sslEngine.setEnableSessionCreation(false); // disable renegotiation
+                    return HandshakeState.COMPLETED;
+                case NEED_TASK:
+                    sslEngine.getDelegatedTask().run();
+                    break;
+                case NEED_UNWRAP:
+                    if (hasWrapBufferMoreData()) return HandshakeState.NEED_WRITE;
+                    int bytesUnwrapped = sslEngineUnwrap(handshakeDummyBuffer);
+                    if (handshakeDummyBuffer.position() > 0) throw new SSLException("Got application data in handshake unwrap: " + handshakeDummyBuffer);
+                    if (bytesUnwrapped == -1) return HandshakeState.NEED_READ;
+                    break;
+                case NEED_WRAP:
+                    int bytesWrapped = sslEngineWrap(handshakeDummyBuffer);
+                    if (bytesWrapped == -1) return HandshakeState.NEED_WRITE;
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected handshake status: " + sslEngine.getHandshakeStatus());
+            }
+        }
+    }
+
+    private static HandshakeResult toHandshakeResult(HandshakeState state) {
+        switch (state) {
+            case NEED_READ:
+                return HandshakeResult.NEED_READ;
+            case NEED_WRITE:
+                return HandshakeResult.NEED_WRITE;
+            case COMPLETED:
+                return HandshakeResult.DONE;
+            default:
+                throw new IllegalStateException("Unhandled state: " + state);
+        }
+    }
+
+    @Override
+    public int getMinimumReadBufferSize() {
+        return sslEngine.getSession().getApplicationBufferSize();
+    }
+
+    @Override
+    public int read(ByteBuffer dst) throws IOException {
+        verifyHandshakeCompleted();
+        int bytesUnwrapped = sslEngineAppDataUnwrap(dst);
+        if (bytesUnwrapped > 0) return bytesUnwrapped;
+
+        int bytesRead = channelRead();
+        if (bytesRead == 0) return 0;
+        return drain(dst);
+    }
+
+    @Override
+    public int drain(ByteBuffer dst) throws IOException {
+        verifyHandshakeCompleted();
+        int totalBytesUnwrapped = 0;
+        int bytesUnwrapped;
+        do {
+            bytesUnwrapped = sslEngineAppDataUnwrap(dst);
+            totalBytesUnwrapped += bytesUnwrapped;
+        } while (bytesUnwrapped > 0);
+        return totalBytesUnwrapped;
+    }
+
+    private int sslEngineAppDataUnwrap(ByteBuffer dst) throws IOException {
+        int bytesUnwrapped = sslEngineUnwrap(dst);
+        if (bytesUnwrapped == 0) throw new SSLException("Got handshake data in application data unwrap");
+        if (bytesUnwrapped == -1) return 0;
```

Comment:
seems a bit inconsistent to map -1 to 0 in unwrap, but not in wrap (for app data)